### PR TITLE
Feedback

### LINKED ISSUE #29: Driver feedback after finished order
- [x] Add feedback form after driver finishes order with inputs: feedback text (textarea), stars with numbers below (1-5)
- [x] Change feedback model and rewrite code appropriately to changes

### PR DESCRIPTION
**Pull request by Pr1sm0**
Change toggle actions to single toggle action, add logic for changing
driver id after click on take button, add author and subject roles to feedback, change driver model,
add update order action, change feedback model

### CODE REVIEW

File: client/src/App.tsx [hunk 1]
```diff
@@ -8,17 +8,17 @@ import AppRouterContainer from './pages/AppRouterContainer';
...
-      </Provider>
-    </BrowserRouter>;
+import Feedback from './pages/Feedback/Feedback';
 
+const App = () => (
+  <BrowserRouter>
+    <Provider store={store}>
+      <ConnectedRouter history={history}>
+        <AppRouterContainer />
+        <Feedback></Feedback>
```
Comment (andrii-kuzmichov) [inline client/src/App.tsx]: Can use  short type like as `<Feedback />`?

File: client/src/actions/feedbackActions.ts [hunk 2]
```diff
@@ -1,9 +1,27 @@
-import { FeedbackAction, FeedbackActionTypes } from '../types/feedbackTypes';
+import {
+  ChangeValueAction,
+  FeedbackActionTypes,
+  FeedbackState,
+  FeedbackValues,
+} from '../types/feedbackTypes';
+import { FeedbackFormI } from '../utils/interfaces';
```
Comment (andrii-kuzmichov) [inline client/src/actions/feedbackActions.ts]: Why this interface type import from utils?

File: client/src/actions/orderActions.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ import {
   ChangeValueAction,
   OrderActionTypes,
   OrderValues,
+  UpdateOrderAction,
 } from '../types/orderTypes';
```
Comment (andrii-kuzmichov) [inline client/src/actions/orderActions.ts]: Why you store interface and action type in the same place?

File: client/src/components/Button/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { FC } from 'react';
+import { Button } from 'react-bootstrap';
+import { ButtonPropsI } from '../../utils/interfaces';
```
Comment (andrii-kuzmichov) [inline client/src/components/Button/Button.tsx]: The same

File: client/src/components/Button/Button.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+export const CustomButton: FC<ButtonPropsI> = ({
```
Comment (andrii-kuzmichov) [inline client/src/components/Button/Button.tsx:5]: What is the benefit of this component?

File: client/src/reducers/feedbackReducer.ts [hunk 6]
```diff
@@ -2,18 +2,23 @@ import { FeedbackAction, FeedbackActionTypes } from '../types/feedbackTypes';
...
 ): FeedbackState => {
-  switch (type) {
-    case FeedbackActionTypes.SHOW_MODAL:
-      return { isShown: true };
-    case FeedbackActionTypes.HIDE_MODAL:
-      return { isShown: false };
+  switch (action.type) {
+    case FeedbackActionTypes.TOGGLE_MODAL:
+      return { ...state, isShown: !state.isShown };
+    case FeedbackActionTypes.CHANGE_VALUE:
```
Comment (andrii-kuzmichov) [inline client/src/reducers/feedbackReducer.ts]: What does this action change?

File: client/src/sagas/feedbackSaga.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+function* createFeedbackWorker(
+  action: CreateFeedbackAction,
+): Generator<StrictEffect, void, any> {
+  try {
+    const currentUserId = yield select(userAuthId);
+    const customerId = yield select(userOrderId);
+    const currentOrderId = yield select(orderId);
+    yield put(changeFeedbackValue('text', action.payload.text));
+    yield put(changeFeedbackValue('rating', action.payload.rating));
+    yield put(changeFeedbackValue('orderId', currentOrderId));
```
Comment (andrii-kuzmichov) [inline client/src/sagas/feedbackSaga.ts]: Why didn't you create one action?

File: client/src/sagas/feedbackSaga.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    if (currentUserId === customerId) {
+      yield put(changeFeedbackValue('author_role', UserRoles.USER));
+      yield put(changeFeedbackValue('subject_role', UserRoles.DRIVER));
+    } else {
+      yield put(changeFeedbackValue('author_role', UserRoles.DRIVER));
+      yield put(changeFeedbackValue('subject_role', UserRoles.USER));
```
Comment (andrii-kuzmichov) [inline client/src/sagas/feedbackSaga.ts]: The same for the twice cases

File: client/src/sagas/feedbackSaga.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    const feedback = yield select(getFeedback);
+    yield call(createFeedback, feedback);
```
Comment (andrii-kuzmichov) [inline client/src/sagas/feedbackSaga.ts:31]: why didn't you generate feedback obj in saga?

File: client/src/sagas/orderSaga.ts [hunk 10]
```diff
@@ -26,6 +27,7 @@ function* makeOrderWorker(): Generator<StrictEffect, void, any> {
     const data = yield call(makeOrder(order, userID));
     if (data.status === 200) {
       yield put(changeOrderValue('id', data.data.id));
+      yield put(changeOrderValue('customer_id', data.data.customer_id));
```
Comment (andrii-kuzmichov) [inline client/src/sagas/orderSaga.ts]: the same that changing value above

File: client/src/types/userTypes.ts [hunk 11]
```diff
@@ -43,9 +43,9 @@ export interface IUser {
   isAuth: boolean;
   hasError: boolean;
   id: number | null;
-  driver_info?: IDriver,
-  isDriver: boolean,
-  authError: string,
+  driver_info?: IDriver;
```
Comment (andrii-kuzmichov) [inline client/src/types/userTypes.ts:46]: Why you use snake case formatting?

File: client/src/App.tsx (hunk 1 above)
Comment (Pr1sm0) [reply client/src/App.tsx]: Resolved

File: client/src/actions/feedbackActions.ts (hunk 2 above)
Comment (Pr1sm0) [reply client/src/actions/feedbackActions.ts]: Resolved

File: client/src/actions/orderActions.ts (hunk 3 above)
Comment (Pr1sm0) [reply client/src/actions/orderActions.ts]: Resolved

File: client/src/components/Button/Button.tsx (hunk 4 above)
Comment (Pr1sm0) [reply client/src/components/Button/Button.tsx]: Resolved

File: client/src/components/Button/Button.tsx (hunk 5 above)
Comment (Pr1sm0) [reply client/src/components/Button/Button.tsx:5]: Resolved

File: client/src/reducers/feedbackReducer.ts (hunk 6 above)
Comment (Pr1sm0) [reply client/src/reducers/feedbackReducer.ts]: Resolved

File: client/src/sagas/feedbackSaga.ts (hunk 7 above)
Comment (Pr1sm0) [reply client/src/sagas/feedbackSaga.ts]: Resolved

File: client/src/sagas/feedbackSaga.ts (hunk 8 above)
Comment (Pr1sm0) [reply client/src/sagas/feedbackSaga.ts]: Resolved

File: client/src/types/userTypes.ts (hunk 11 above)
Comment (Pr1sm0) [reply client/src/types/userTypes.ts:46]: Resolved (will be changed in the next pull requests)

File: client/src/sagas/feedbackSaga.ts (hunk 9 above)
Comment (Pr1sm0) [reply client/src/sagas/feedbackSaga.ts:31]: Resolved

File: client/src/sagas/orderSaga.ts (hunk 10 above)
Comment (Pr1sm0) [reply client/src/sagas/orderSaga.ts]: Resolved

**Review by N-Balinskiy — APPROVED** [review]

**Review by romanh24 — APPROVED** [review]

**Review by babchukbohdan — APPROVED** [review]

**Review by andrii-kuzmichov — APPROVED** [review]

**Pull request merged**